Sources/NIO/Selector.swift

```diff
@@ -372,7 +372,21 @@ internal class Selector<R: Registration> {
         case .now:
             return timespec(tv_sec: 0, tv_nsec: 0)
         case .blockUntilTimeout(let nanoseconds):
-            return timespec(timeAmount: nanoseconds)
+            var ts = timespec(timeAmount: nanoseconds)
+            // Check that the timespec conforms to the definition in the C11 standard (ISO/IEC 9899:2011)
+            assert((0..<1_000_000_000).contains(ts.tv_nsec), "\(ts) is malformed (0 <= tv_nsec < 1_000_000_000)")
+
+            // The maximum value in seconds supported by the Darwin-kernel interval timer (kern_time.c:itimerfix())
+            // (note that - whilst unlikely - this value *could* change).
+            let sysIntervalTimerMaxSec = 100_000_000
+
+            // Clamp the timespec tv_sec value to the maximum supported by the Darwin-kernel,
+            // (note this schedules the event many years into the future).
+            if !(0...sysIntervalTimerMaxSec).contains(ts.tv_sec) {
```

Comment:
Hmm, sorry for flipflopping here, but I actually think we should program this defensively. While the assert above prevents us for scheduling a negative timeout, we probably _shouldn't_ turn a negative timeout into a massive positive one. This turns EINVAL (which will alert us to an issue) into a very long delay, which seems not good. So let's only check the upper bound.

Reply:
Let me look at that agains. tv_sec is a time_t which if I recall correctly has some slightly weird wishy washy definition like its a integral value. I suspect it could be negative but for most practical cases won't be.

Reply:
To be clear, the kernel will reject a negative value here. That isn't going to prevent us from _creating_ a negative value. That's the case I want to defend against.

> However, it feels like it is somewhat back to square one as we're back to letting the kernel decide what is and what isn't valid.

I disagree. The two cases have different causes.

The thing we're defending against is a user scheduling a task in the far future causing a crash. This should not happen: there is nothing wrong with scheduling a task in the far future. The kernel will reject it if we simply pass the naive value of that timeout in, so we need to tolerate that error.

Similarly, the user should be able to schedule a task to occur in the past without causing a crash! _However_, the code elsewhere in the codebase handles that conversion: we should not be creating negative timeouts by the time we get here. If we see a crash this is not _user_ error, this is our programming error. That's why its acceptable to crash here on negative values.

Putting it another way, the Linux and Darwin kernels behave differently here. This code is responsible for compatibility between the two platforms. They do not differ on negative timeout values, and so we ensure we get that right on all platforms.